File: xls-sql-converter/src/main/java/ru/yandex/market/v1/table/Field.java
package ru.yandex.market.v1.table;

public class Field {
    private static final String FORMAT = "'%s'";
    private static final String N_FORMAT = "N" + FORMAT;
    private final String rawValue;
    private final boolean quoted;

    public Field(String rawValue, boolean quoted) {
        this.rawValue = rawValue;
        this.quoted = quoted;
    }

    public static Field of(int value) {
        return of(String.valueOf(value));
    }

    public static Field of(double value) {
        return of(String.valueOf(value));
    }

    public static Field of(String value) {
        return new Field(value, false);
    }

    public static Field quoted(String value) {
        return new Field(value, true);
    }

    public String getRawValue() {
        return rawValue;
    }

    public String getSqlValue() {
        if (quoted) {
            String format = rawValue.codePoints().anyMatch(n -> n >= Byte.MAX_VALUE) ? N_FORMAT : FORMAT;
            return String.format(format, rawValue);
        } else {
            return rawValue;
        }
    }

    @Override
    public String toString() {
        return rawValue;
    }
}

File: xls-sql-converter2/src/main/kotlin/app/converter/table/Field.kt
package app.converter.table

class Field(val rawValue: String, quoted: Boolean) {
    companion object {
        fun of(value: Int): Field = of(value.toString())
        fun of(value: Double): Field = of(value.toString())
        fun of(value: Boolean): Field = of(if (value) 1 else 0)
        fun of(value: String): Field = Field(value, false)
        fun quoted(value: String): Field = Field(value, true)
    }

    val sqlValue: String =
        if (quoted) {
            val prefix = if (rawValue.codePoints().anyMatch { it >= Byte.MAX_VALUE }) "N" else ""
            "$prefix'$rawValue'"
        } else rawValue

}
